File: Online Judges/LeetCode/June LeetCoding Challenge/Week-3/Dungeon Game.cpp
class Solution
{
public:
    int calculateMinimumHP(vector<vector<int>> &dungeon)
    {
        int n = dungeon.size();
        int m = dungeon[0].size();
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, INT_MAX));
        dp[n][m - 1] = 1;
        dp[n - 1][m] = 1;
        for (int i = n - 1; i >= 0; i--)
        {
            for (int j = m - 1; j >= 0; j--)
            {
                dp[i][j] = min(dp[i][j + 1], dp[i + 1][j]) - dungeon[i][j];
                dp[i][j] = (dp[i][j] <= 0) ? 1 : dp[i][j];
            }
        }
        return dp[0][0];
    }
};

File: Online Judges/LeetCode/June LeetCoding Challenge/Week-3/Dungeon Game.rs
use std::cmp::{min, max};
impl Solution {
    pub fn calculate_minimum_hp(dungeon: Vec<Vec<i32>>) -> i32 {
        let n = dungeon.len();
        let m = dungeon[0].len();
        
        let mut dp:Vec<Vec<i32>> = vec![vec![std::i32::MAX; m+1];n+1];
        
        dp[n-1][m] = 1;
        dp[n][m-1] = 1;
        
        for i in (0..n).rev() {
            for j in (0..m).rev() {
                dp[i][j] = max(min(dp[i][j+1], dp[i+1][j]) - dungeon[i][j], 1);
            }
        }
        dp[0][0]
    }
}